rtpsender.go

```diff
@@ -1,154 +1,143 @@
-	r := &RTPSender{
-		Track:     track,
-		transport: transport,
-		api:       api,
+func (api *API) NewRTPSender(track *Track, transport *DTLSTransport) (*RTPSender, error) {
+	if track == nil {
+		return nil, fmt.Errorf("Track must not be nil")
+	} else if transport == nil {
+		return nil, fmt.Errorf("DTLSTransport must not be nil")
 	}
 
-	r.Track.sampleInput = make(chan media.Sample, 15) // Is the buffering needed?
-	r.Track.rawInput = make(chan *rtp.Packet, 15)     // Is the buffering needed?
-	r.Track.rtcpInput = make(chan rtcp.Packet, 15)    // Is the buffering needed?
-
-	r.Track.Samples = r.Track.sampleInput
-	r.Track.RawRTP = r.Track.rawInput
-	r.Track.RTCPPackets = r.Track.rtcpInput
-
-	if r.Track.isRawRTP {
-		close(r.Track.Samples)
-	} else {
-		close(r.Track.RawRTP)
+	track.mu.RLock()
+	defer track.mu.RUnlock()
+	if track.receiver != nil {
+		return nil, fmt.Errorf("RTPSender can not be constructed with remote track")
 	}
 
-	return r
+	return &RTPSender{
+		track:     track,
+		transport: transport,
+		api:       api,
+	}, nil
 }
 
 // Send Attempts to set the parameters controlling the sending of media.
-func (r *RTPSender) Send(parameters RTPSendParameters) {
-	if r.Track.isRawRTP {
-		go r.handleRawRTP(r.Track.rawInput)
-	} else {
-		go r.handleSampleRTP(r.Track.sampleInput)
-	}
-
-	go r.handleRTCP(r.transport, r.Track.rtcpInput)
-}
+func (r *RTPSender) Send(parameters RTPSendParameters) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
 
-// Stop irreversibly stops the RTPSender
-func (r *RTPSender) Stop() {
-	if r.Track.isRawRTP {
-		close(r.Track.RawRTP)
-	} else {
-		close(r.Track.Samples)
+	if r.sendCalled {
+		return fmt.Errorf("Send has already been called")
 	}
+	r.sendCalled = true
 
-	// TODO properly tear down all loops (and test that)
-}
+	srtcpSession, err := r.transport.getSRTCPSession()
+	if err != nil {
+		return err
+	}
+	srtcpReadStream, err := srtcpSession.OpenReadStream(parameters.encodings.SSRC)
+	if err != nil {
+		return err
+	}
+	r.rtcpReadStream = newLossyReadCloser(srtcpReadStream)
 
-func (r *RTPSender) handleRawRTP(rtpPackets chan *rtp.Packet) {
-	for {
-		p, ok := <-rtpPackets
-		if !ok {
-			return
-		}
+	r.track.mu.Lock()
+	r.track.senders = append(r.track.senders, r)
+	r.track.mu.Unlock()
 
-		r.sendRTP(p)
-	}
+	return nil
 }
 
-func (r *RTPSender) handleSampleRTP(rtpPackets chan media.Sample) {
-	packetizer := rtp.NewPacketizer(
-		rtpOutboundMTU,
-		r.Track.PayloadType,
-		r.Track.SSRC,
-		r.Track.Codec.Payloader,
-		rtp.NewRandomSequencer(),
-		r.Track.Codec.ClockRate,
-	)
-
-	for {
-		in, ok := <-rtpPackets
-		if !ok {
-			return
-		}
-		packets := packetizer.Packetize(in.Data, in.Samples)
-		for _, p := range packets {
-			r.sendRTP(p)
+// Stop irreversibly stops the RTPSender
+func (r *RTPSender) Stop() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stopCalled {
+		return fmt.Errorf("Stop has already been called")
```

Comment:
If Stop() was already called is it really an error to call it again? Could just exit early. Likewise for Close()...

Reply:
Agree! Fixed with https://github.com/pions/webrtc/pull/437/commits/a40fc1f385f17a35285edf720bfa33c9ac9fe243